proxy/proxy/client.go

```diff
@@ -225,6 +228,15 @@ func (c *Client) Dial(instance string) (net.Conn, error) {
 	if err != nil {
 		return nil, err
 	}
+
+	// Refresh cert 5 minutes before it expires.
+	timer := time.NewTimer(cfg.Certificates[0].Leaf.NotAfter.Sub(time.Now()) - refreshCertBuffer)
+	go func() {
+		<-timer.C
+		logging.Verbosef("Cert for instance %s will expire soon, refreshing now.", instance)
+		c.refreshCfg(instance)
```

Comment:
Due to the way refreshCfg/cachedCfg works (refreshCfg grabs the write-lock of `c.cfgL`, and cachedCfg grabs the read-lock), calling this in a background goroutine is not sufficient to prevent delays of Dialing. Previously the delay would happen during the first dials after the cert expires. Now, any dials that happen 5 minutes before the cert expires will block just the same. You'll have to do something fancier with the locks to get concurrent Dials during the time refreshCfg is contacting the backend.

Reply:
Ah, I see what you mean now. Is there any reason to hold the lock for the entirety of the method? What about only using the read lock to get the instance data from cfgCache and only lock when updating cfgCache in the deferred function? Sorry if it's a dumb question, still trying to get up to speed on the code.

Reply:
No dumb questions, if the code is hard to understand I'm the one to blame :)

The write lock is being held the entire call to refreshConfig to prevent concurrent dials. For example, if two application threads try to open a new connection to a database which has expired config at the same time, we only want to contact the Cloud SQL API on one of the corresponding goroutines while the other waits for the cert to be updated. Right now the lock is very coarse-grained: only one goroutine is allowed to contact the API at a time and no goroutines are allowed to read the config during this period. It was designed this way because we figured that most of the time there was one or two instances that the proxy was going to connect to, so this simpler, coarser locking structure was advantageous to code clarity without really affecting performance (since other goroutines probably had to wait for their cert to be updated anyway).

As you suggest, I agree from a thread-safety point of view it is safe to just unlock the write lock during the calls to `c.Certs.Local` and `c.Certs.Remote`, but then those two threads would try to contact the backend at the same time instead of redialing on their own. Since there are potentially hundreds of application threads that could try to do a redial at the same time I don't think it will scale very well. I think we need to add a bit more depth to the locking strategy. Here's some requirements I can think of:

   1) Readers must be able to read cached config while goroutines are contacting the API
   2) Only one goroutine may be contacting backend at a time (at least, per instance)
   3) When a reader finds the cached config is invalid/expired AND there is already a goroutine contacting the backend, this reader should wait for and use the response that goroutine gets (rather than contacting the backend itself).

Given those requirements I whipped up something that I think could work, but you should definitely look it over to make sure it isn't witchcraft (i.e. I haven't even compiled it). And if you have another implementation you can think of that you like more free to use it. You can ping me over chat/email if you need more clarification or anything.

    type Client struct {
       // A bunch of other stuff
       cfgCache map[string]*cacheEntry
       cfgL     sync.RWMutex
    }

    type cacheEntry struct {
        timestamp time.Time
        // The err, addr, and cfg fields of this entry must not be read before this channel has been closed.
        ready chan struct{}
        // If err is not nil, the addr and cfg are not valid.
        err  error
        addr string
        cfg  *tls.Config
     }

    func (c *Client) refreshCfg(instance string) (addr string, cfg *tls.Config, err error) {
       throttle := c.RefreshCfgThrottle
       if throttle == 0 {
          throttle = DefaultRefreshCfgThrottle
       }

       c.cfgL.Lock()

       // cfgCache is lazy-init
       if c.cfgCache == nil {
          c.cfgCache = make(map[string]*cacheEntry)
       }

       old, oldok := c.cfgCache[instance]

       if oldok && time.Since(old.timestamp) < throttle {
          c.cfgL.Unlock()
          logging.Errorf("Throttling refreshCfg(%s): it was only called %v ago", instance, time.Since(old.lastRefreshed))
          // Refresh was called too recently, just reuse the result after it is ready
          <-old.ready
          return old.addr, old.cfg, old.err
       }

       ent := &cacheEntry{
          timestamp: time.Now(),
          ready: make(chan struct{}),
       }
       c.cfgCache[instance] = ent
       c.cfgL.Unlock()

       defer func() {
          ent.err = err
          ent.addr = addr
          ent.cfg = cfg
          // Now that we've written the response in the entry, close the channel to
          // signal that the cache entry is ready.
          close(ent.ready)
       }()

       mycert, err := c.Certs.Local(instance)
       if err != nil {
          return "", nil, err
       }

       scert, addr, name, err := c.Certs.Remote(instance)
       if err != nil {
          return "", nil, err
       }
       certs := x509.NewCertPool()
       certs.AddCert(scert)

       cfg = &tls.Config{
          ServerName:   name,
          Certificates: []tls.Certificate{mycert},
          RootCAs:      certs,
       }
       return fmt.Sprintf("%s:%d", addr, c.Port), cfg, nil
    }

Reply:
What about using two separate locks: one to protect the number of concurrent API calls and one to protect the cache? The former would be held through the entirety of refreshCfg to allow for only one goroutine to make the API calls at a time, and the latter would allow other goroutines to access the cache while the cert is refreshed.

I don't think this would meet criteria number 3, but if the idea is to refresh the cert a few minutes before it expires, I don't think readers will typically need to wait for the cert to be refreshed.

PTAL at my changes. I feel like I'm probably missing something. :)